### PR TITLE
mysql: used brewed dependencies

### LINKED ISSUE #72333: Missing info in mysql_config prevents building/linking against it
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

$ brew config
HOMEBREW_VERSION: 3.0.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: ee52b19179d056f3dd2fdf9c52214d3d51dbb9d9
Last commit: 87 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 26105415394a70964740cb9f671d1634c196b9ea
Core tap last commit: 66 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /usr/bin/vim
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit skylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.32.29
Xcode: 12.4
$ brew doctor
Your system is ready to brew.

### 

- [X] The `brew doctor` above contains no "Warning" lines.

### What were you trying to do (and why)?

Build DBD::mysql from the CPAN (because I need to talk to a MySQL database from my perl code)

### What happened (include all command output)?

The build failed, because "Can't link/include C library 'ssl', 'crypto', aborting". I won't provide all the output because it goes on for pages and pages and is mostly irrelevant. The relevant bit is ...

```
I will use the following settings for compiling and testing:

  cflags        (mysql_config) = -I/usr/local/Cellar/mysql/8.0.23_1/include/mysql
  embedded      (guessed     ) =
  ldflags       (guessed     ) =
  libs          (mysql_config) = -L/usr/local/Cellar/mysql/8.0.23_1/lib -lmysqlclient -lssl -lcrypto
-lresolv
  mysql_config  (guessed     ) = mysql_config
  nocatchstderr (default     ) = 0
  nofoundrows   (default     ) = 0
  nossl         (default     ) = 0
  testdb        (default     ) = test
  testhost      (default     ) =
  testpassword  (default     ) =
  testport      (default     ) =
  testsocket    (default     ) =
  testuser      (guessed     ) = dcantrell

To change these settings, see 'perl Makefile.PL --help' and
'perldoc DBD::mysql::INSTALL'.

Checking if libs are available for compiling...
Can't link/include C library 'ssl', 'crypto', aborting.
-> N/A
-> FAIL Configure failed for DBD-mysql-4.050. See /Users/dcantrell/.cpanm/work/1614703692.86342/build
.log for details.
```

### What did you expect to happen?

The build to succeed. To *make* it succeed, edit `/usr/local/bin/mysql_config`, which is installed by the formula, and immediately after this line:

```
libs="-L$pkglibdir"
```

insert this line:

```
libs="$libs -L/usr/local/opt/openssl/lib"
```

### Step-by-step reproduction instructions (by running `brew` commands)

$ brew install mysql # or reinstall
$ cpanm DBD::mysql

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

~~It was suggested upstream that copying over the libraries when a path is~~
~~passed to `WITH_SSL` is intended behaviour. If we don't want these~~
~~libraries to be copied, we should be setting `WITH_SSL=system`.~~

~~See https://bugs.mysql.com/bug.php?id=103227.~~

So that was a lie. Let's just fix this to vendor less stuff instead.

### CODE REVIEW

**Comment by carlocab** [comment]
Marking this as a draft with `do not merge` for now, because I'll want to inspect the bottle first before merging.

**Comment by cho-m** [comment]
@carlocab do you want to look into the other potentially vendored dependencies
(e.g. `ICU`, `LZ4`, `CURL`[though curl is probably not enabled here]).

Maybe just switching to mega `cmake` flag `WITH_SYSTEM_LIBS`
(though not sure how it impacts disabled options like curl).

**Comment by carlocab** [comment]
Yea, I'll probably turn this PR into that. Had a peek at the ARM build, and it still copied the OpenSSL libraries into `lib`, so the workaround is still needed.

Using brewed ICU is a bit of a pain though since ICU always breaks the API. (See any `icu4c` version bump PR.)

**Comment by carlocab** [comment]
Do we want to disable `curl`? We get it for free on macOS. It might be worth enabling it only on macOS, unless there's a good reason not to?

**Comment by cho-m** [comment]
Not sure. It was disable on previous bump PR, but I didn't check if this was default behavior and what exact features cURL enabled in mySQL.

**Comment by cho-m** [comment]
MySQL at least enables `curl` for their commercial builds, though I still didn't check if the curl-related features are open-source: https://github.com/mysql/mysql-server/commit/f18099b5e6b4a4db9d371d6fda6ed81437c5aa8a

Since it is available on macOS, I don't think there is any harm in enabling.

**Comment by carlocab** [comment]
Agreed. Do you have a link to the commit where it was disabled?

**Comment by carlocab** [comment]
Gentoo and FreeBSD build with `WITH_CURL=system`, so I think enabling it should be fine.

**Comment by carlocab** [comment]
Checked the Big Sur bottle. Can't find linkage with `curl`. Either it doesn't use `libcurl` (perhaps just a `curl` executable?), or it goes out of its way to avoid linkage with the system `libcurl` on macOS.

**Comment by carlocab** [comment]
Let's try brewed `curl`. Though, I am unsure about one thing: if it can't find a system lib, does it build its own, or does it just not build the functionality enabled by the library?

If it's the former we want to add these deps. If it's the latter, then we can do without them as long as no one is asking for them.

**Comment by carlocab** [comment]
Ok, `lz4` is bundled: https://github.com/mysql/mysql-server/blob/98b2ccb470de120d36bc4a623c814cdfded958ec/cmake/lz4.cmake#L23-L24

So is ICU: https://github.com/mysql/mysql-server/blob/98b2ccb470de120d36bc4a623c814cdfded958ec/cmake/icu.cmake#L31

`curl` is not bundled: https://github.com/mysql/mysql-server/blob/98b2ccb470de120d36bc4a623c814cdfded958ec/cmake/curl.cmake#L23-L25

**Comment by carlocab** [comment]
Upstream have acknowledged the bug we work around here, which is progress: https://bugs.mysql.com/bug.php?id=103227

**Review by nandahkrishna — APPROVED** [review]

**Pull request closed without merging**
